lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,274 @@
+// sorted iteration of its element.
+type memorySorted struct {
+	bytes    int
+	outgoing bool
+	values   []Value
+	keys     [][]byte
+}
+
+func (o *memorySorted) add(k []byte, v Value) {
+	if o.outgoing {
+		panic("Add/Append may never be called after PopFirst/PopLast")
+	}
+	o.values = append(o.values, v)
+	o.keys = append(o.keys, k)
+	o.bytes += v.ProtoSize()
+}
+
+func (o *memorySorted) Len() int {
+	return len(o.values)
+}
+func (o *memorySorted) Swap(a, b int) {
+	o.values[a], o.values[b] = o.values[b], o.values[a]
+	o.keys[a], o.keys[b] = o.keys[b], o.keys[a]
+}
+func (o *memorySorted) Less(a, b int) bool {
+	return bytes.Compare(o.keys[a], o.keys[b]) == -1
+}
+
+func (o *memorySorted) Bytes() int {
+	return o.bytes
+}
+
+func (o *memorySorted) Close() {}
+
+func (o *memorySorted) Items() int {
+	return len(o.values)
+}
+
+func (o *memorySorted) PopFirst(v Value) bool {
+	return o.pop(v, true)
+}
+
+func (o *memorySorted) PopLast(v Value) bool {
+	return o.pop(v, false)
+}
+
+func (o *memorySorted) pop(v Value, first bool) bool {
+	if !o.outgoing {
+		sort.Sort(o)
+		o.outgoing = true
+	}
+	if o.Items() == 0 {
+		return false
+	}
+	if first {
+		copyValue(v, o.values[0])
+		o.values = o.values[1:]
+		o.keys = o.keys[1:]
+	} else {
+		i := o.Items() - 1
+		copyValue(v, o.values[i])
+		o.values = o.values[:i]
+		o.keys = o.keys[:i]
+	}
+	o.bytes -= v.ProtoSize()
+	return true
+}
+
+func (o *memorySorted) newIterator(parent iteratorParent, reverse bool) keyIterator {
+	if !o.outgoing {
+		sort.Sort(o)
+	}
+	return &memSortIterator{
+		memIterator: newMemIterator(o.values, parent, reverse, o.Items()),
+		keys:        o.keys,
+	}
+}
+
+type memSortIterator struct {
+	*memIterator
+	keys [][]byte
+}
+
+func (si *memSortIterator) key() []byte {
+	if si.pos == si.len || si.pos == -1 {
+		return nil
+	}
+	return si.keys[si.pos]
+}
+
+// diskSorted is backed by a LevelDB database in a temporary directory. It relies
+// on the fact that iterating over the database is done in key order.
+type diskSorted struct {
+	*diskMap
+	bytes int
+}
+
+func newDiskSorted(loc string) *diskSorted {
+	return &diskSorted{diskMap: newDiskMap(loc)}
+}
+
+func (o *diskSorted) add(k []byte, v Value) {
+	suffix := make([]byte, suffixLength)
+	binary.BigEndian.PutUint64(suffix[:], uint64(o.Items()))
+	o.diskMap.addBytes(append(k, suffix...), v)
+	o.bytes += v.ProtoSize()
+}
+
+func (o *diskSorted) Bytes() int {
+	return o.bytes
+}
+
+func (o *diskSorted) PopFirst(v Value) bool {
+	return o.pop(v, true)
+}
+
+func (o *diskSorted) PopLast(v Value) bool {
+	return o.pop(v, false)
+}
+
+func (o *diskSorted) pop(v Value, first bool) bool {
+	it := o.db.NewIterator(nil, nil)
+	defer it.Release()
+	var ok bool
+	if first {
+		ok = it.First()
+	} else {
+		ok = it.Last()
+	}
+	if !ok {
+		return false
+	}
+	errPanic(v.Unmarshal(it.Value()))
+	errPanic(o.db.Delete(it.Key(), nil))
+	o.bytes -= v.ProtoSize()
+	o.len--
+	return true
+}
```

Comment:
Perhaps allocating an iterator is cheap but the implementation seems a bit silly.
Can't we track the index of head and tail as we pop first pop last?

Reply:
I don't see any way to get the key or element of the first/last entry in goleveldb other than through an iterator. We could keep the iterator open though, so it can serve multiple `pop`s.